### PR TITLE
Issue #714 Added the fuzzy test.

### PR DESCRIPTION
**Pull request by monadbobo**
sumarry:

1. Added a new parameter "fuzzer" to nebula_add_test(cmake).
   Turn on this parameter indicates that the current test is a fuzzy test.
2. Currently only a simple fuzzy test has been added for paser module.
3. The current fuzzy test only supports Clang, so if the compiler was gcc it will skip.

### CODE REVIEW

**Comment by monadbobo** [comment]
Jenkins go!

**Comment by dutor** [comment]
I didn't find any official documentation  to describe how this sanitizer works, did you?

**Comment by monadbobo** [comment]
> I didn't find any official documentation to describe how this sanitizer works, did you?

http://llvm.org/docs/LibFuzzer.html

File: CMakeLists.txt [hunk 1]
```diff
@@ -355,28 +355,41 @@ macro(nebula_add_test)
         nebula_test                 # prefix
         "DISABLED"                  # <options>
         "NAME"                      # <one_value_args>
-        "SOURCES;OBJECTS;LIBRARIES" # <multi_value_args>
+        "SOURCES;OBJECTS;LIBRARIES;FUZZER" # <multi_value_args>
```
Comment (dutor) [inline CMakeLists.txt]: IMO, `FUZZER` is an option which carries no arguments, like `DISABLED` above. So you should place it in the _options_ list.

File: CMakeLists.txt [hunk 2]
```diff
@@ -355,28 +355,41 @@ macro(nebula_add_test)
...
-        string(REGEX REPLACE "${CMAKE_SOURCE_DIR}/src/(.*)/test" "\\1" test_group ${CMAKE_CURRENT_SOURCE_DIR})
-        add_test(NAME ${nebula_test_NAME} COMMAND ${nebula_test_NAME})
-        set_tests_properties(${nebula_test_NAME} PROPERTIES LABELS ${test_group})
-        # e.g. cmake -DNEBULA_ASAN_PRELOAD=/path/to/libasan.so
-        # or,  cmake -DNEBULA_ASAN_PRELOAD=`/path/to/gcc --print-file-name=libasan.so`
-        if (NEBULA_ASAN_PRELOAD)
-            set_property(
-                TEST ${nebula_test_NAME}
-                PROPERTY ENVIRONMENT LD_PRELOAD=${NEBULA_ASAN_PRELOAD}
+    if (${nebula_test_FUZZER})
```
Comment (dutor) [inline CMakeLists.txt]: How about changing like this, which seems more cleaner:
```cmake
...
nebula_add_executable(
...
)

if (${nebula_test_FUZZER})
    if (${CMAKE_CXX_COMPILER_ID} STREQUAL "Clang")
        set_target_properties(${nebula_test_NAME} PROPERTIES COMPILE_FLAGS "-g -fsanitize=fuzzer")
        set_target_properties(${nebula_test_NAME} PROPERTIES LINK_FLAGS "-fsanitize=fuzzer")
    endif()
endif()

if (NOT ${nebula_test_DISABLED})
...
```

File: CMakeLists.txt (hunk 1 above)
Comment (dutor) [reply CMakeLists.txt]: Well, given a second thought, `FUZZER ON/OFF` is more expressive.

**Comment by dangleptr** [comment]
Awesome!

**Pull request merged**